Implement multi-chain OCR3 Keybundle

vyzaldysanchez opened the pull request.

vyzaldysanchez committed:
Adds `onchainSigningStrategy` job spec field support

vyzaldysanchez committed:
Makes changeset internal

vyzaldysanchez committed:
Fixes typo and remove unnecessary logs

vyzaldysanchez committed:
Merge branch 'develop' into task/BCF-3136/generic-public-onchain-signing-stragegy

vyzaldysanchez committed:
Stores `onchain_signing_strategy` values in the DB

vyzaldysanchez committed:
Merge branch 'develop' into task/BCF-3136/generic-public-onchain-signing-stragegy

vyzaldysanchez committed:
Fixes test error

vyzaldysanchez committed:
Implements OCR3 multi-chain keybundle adapter WIP

vyzaldysanchez committed:
Implements OCR3 multi-chain keybundle adapter WIP

vyzaldysanchez committed:
Merge branch 'develop' into task/BCF-3136/generic-public-onchain-signing-stragegy

vyzaldysanchez committed:
Fixes merge conflicts

cedric-cordenier commented on core/services/ocrcommon/adapters.go:
```diff
@@ -71,3 +76,66 @@ func (c *OCR3ContractTransmitterAdapter) Transmit(ctx context.Context, digest oc
+	}, ri.Report, signature)
+}
+
+func (a *OCR3OnchainKeyringMultiChainAdapter) MaxSignatureLength() int {
```
@vyzaldysanchez You can iterate over all of the underlying keybundles and return the maximum.

cedric-cordenier commented on core/services/ocrcommon/adapters.go:
```diff
@@ -71,3 +76,66 @@ func (c *OCR3ContractTransmitterAdapter) Transmit(ctx context.Context, digest oc
+	}, r.Report)
+}
+
+func (a *OCR3OnchainKeyringMultiChainAdapter) Verify(opk ocrtypes.OnchainPublicKey, digest ocrtypes.ConfigDigest, seqNr uint64, ri ocr3types.ReportWithInfo[[]byte], signature []byte) bool {
```
@vyzaldysanchez If you unmarshal ri into a structpb.Struct and then extract a specific key from it, that should do what you want, eg. (keyBundleIdentifier = types.NetworkEVM).

```
info := structpb.Struct{}
err := proto.Unmarshal(ri.Info, &info)
if err != nil { // handle }

infoMap := info.AsMap // map[string]any
keyBundleId = infoMap["keyBundleIdentifier"]

selected, ok := a.keyBundles(keyBundleId)
if !ok { // handle incorrect identifier }

return selected.Sign...
```

cedric-cordenier replied:
You can do the same thing for Verify and for Sign

cedric-cordenier commented on core/services/ocrcommon/adapters.go:
```diff
@@ -71,3 +76,66 @@ func (c *OCR3ContractTransmitterAdapter) Transmit(ctx context.Context, digest oc
+	publicKey  ocrtypes.OnchainPublicKey
+}
+
+func NewOCR3OnchainKeyringMultiChainAdapter(ks keystore.OCR2, st ocr2.OCR2OnchainSigningStrategy) (*OCR3OnchainKeyringMultiChainAdapter, error) {
```
@vyzaldysanchez This is pretty good! But I would loosen the coupling a bit here -- the config should provide a mapping of names -> keyBundleIds (which we've validated are valid), and that's it.

We shouldn't depend on eg. types.SupportedRelays since that will break existing jobs if we add a new relay without changing updating the job config.

vyzaldysanchez committed:
Fixes linter

vyzaldysanchez committed:
Improves OCR3 key bundle adapter

vyzaldysanchez committed:
Fixes import cycle

vyzaldysanchez committed:
Fixes import cycle

vyzaldysanchez committed:
Merge branch 'develop' into task/BCF-3136/generic-public-onchain-signing-stragegy

vyzaldysanchez committed:
Fixes lint

vyzaldysanchez committed:
Fixes lint

vyzaldysanchez committed:
Merge branch 'develop' into task/BCF-3136/generic-public-onchain-signing-stragegy

vyzaldysanchez committed:
Adds tests for new adapter

vyzaldysanchez committed:
[[ -s "/Users/vyzaldysanchez/.gvm/scripts/gvm" ]] && source "/Users/vyzaldysanchez/.gvm/scripts/gvm"
Merge remote-tracking branch 'origin/develop' into task/BCF-3136/generic-public-onchain-signing-stragegy

vyzaldysanchez committed:
Updates go.mod

vyzaldysanchez committed:
Merge branch 'develop' into task/BCF-3136/generic-public-onchain-signing-stragegy

vyzaldysanchez committed:
Fixes build

vyzaldysanchez committed:
Fixes lint

vyzaldysanchez committed:
Uses proper `proto` marshall/unmarshall

vyzaldysanchez committed:
Merge branch 'develop' into task/BCF-3136/generic-public-onchain-signing-stragegy

krehermann commented on core/services/ocrcommon/adapters_test.go:
```diff
@@ -2,15 +2,25 @@ package ocrcommon_test
 	"reflect"
 	"testing"
 
+	structpb "github.com/golang/protobuf/ptypes/struct"
```
this package is deprecated. i think is part of the reason you were struggle to make a test.

see https://pkg.go.dev/github.com/golang/protobuf/ptypes/struct

` Deprecated: Use the "google.golang.org/protobuf" module instead.`

that pkg has a NewStruct func that is easier to use https://pkg.go.dev/google.golang.org/protobuf/types/known/structpb?utm_source=godoc#NewStruct

krehermann commented on core/services/ocrcommon/adapters_test.go:
```diff
@@ -105,6 +115,60 @@ func TestOCR3OnchainKeyringAdapter(t *testing.T) {
+	reportInfo := ocr3types.ReportWithInfo[[]byte]{
+		Report: []byte("multi-chain-report"),
+	}
+	info := structpb.Struct{
```
example using the new pkg https://go.dev/play/p/XeVfKrGookX

krehermann reviewed (dismissed):
please update the struct pb pkg in all the call sites and then update the test

george-dorin commented on core/services/ocrcommon/adapters.go:
```diff
@@ -71,3 +80,103 @@ func (c *OCR3ContractTransmitterAdapter) Transmit(ctx context.Context, digest oc
+		a.lggr.Warn("Verify: keyBundleName is not a string")
+		return false
+	}
+	kb := a.keyBundles[name]
```
Should you not check if `a.keyBundles[name]` exists?

vyzaldysanchez committed:
Uses correct proto package

vyzaldysanchez committed:
Merge branch 'develop' into task/BCF-3136/generic-public-onchain-signing-stragegy

vyzaldysanchez committed:
Fixes gomod

vyzaldysanchez committed:
Merge branch 'develop' into task/BCF-3136/generic-public-onchain-signing-stragegy

vyzaldysanchez committed:
Merge branch 'develop' into task/BCF-3136/generic-public-onchain-signing-stragegy

vyzaldysanchez committed:
Merge branch 'develop' into task/BCF-3136/generic-public-onchain-signing-stragegy

cedric-cordenier commented on core/services/ocrcommon/adapters.go:
```diff
@@ -71,3 +80,110 @@ func (c *OCR3ContractTransmitterAdapter) Transmit(ctx context.Context, digest oc
+}
+
+func (a *OCR3OnchainKeyringMultiChainAdapter) Sign(digest ocrtypes.ConfigDigest, seqNr uint64, r ocr3types.ReportWithInfo[[]byte]) (signature []byte, err error) {
+	info := new(structpb.Struct)
```
@vyzaldysanchez You have a little bit of duplicated code here to handle unmarshaling the info and extracting the value you need. Can we extract this out to a private getter?

cedric-cordenier commented on core/services/ocrcommon/adapters.go:
```diff
@@ -71,3 +80,110 @@ func (c *OCR3ContractTransmitterAdapter) Transmit(ctx context.Context, digest oc
+
+var _ ocr3types.OnchainKeyring[[]byte] = (*OCR3OnchainKeyringMultiChainAdapter)(nil)
+
+type OnchainSigningStrategy interface {
```
@vyzaldysanchez Can we avoid defining this as an interface? I'm worried about the coupling this introduces on lines 98-109: you're passing in the strategy and then immediately calling two methods on it.

Can we just pass in a map[string]ocr2key.KeyBundle directly, having extracted the relevant config beforehand? This will result in a bit more boiler plate in the original callsite, but a cleaner implementation overall.

cedric-cordenier commented on core/services/ocr2/delegate.go:
```diff
@@ -714,6 +718,15 @@ func (d *Delegate) newServicesGenericPlugin(
 		}
+		var onchainKeyringAdapter ocr3types.OnchainKeyring[[]byte]
+		if onchainSigningStrategy.IsMultiChain() {
+			onchainKeyringAdapter, err = ocrcommon.NewOCR3OnchainKeyringMultiChainAdapter(d.ks, &onchainSigningStrategy, lggr)
```
@vyzaldysanchez We should extract the config here and transform it into a map[string]ocr2KeyBundle before passing it into the multichain adapter.

vyzaldysanchez committed:
Improves implementation

vyzaldysanchez committed:
Merge branch 'develop' into task/BCF-3136/generic-public-onchain-signing-stragegy

vyzaldysanchez committed:
Merge branch 'develop' into task/BCF-3136/generic-public-onchain-signing-stragegy

vyzaldysanchez committed:
Fixes validation on optional setup

vyzaldysanchez committed:
Merge branch 'develop' into task/BCF-3136/generic-public-onchain-signing-stragegy
